flink-runtime/src/main/java/org/apache/flink/runtime/rest/FileUploadHandler.java

```diff
@@ -134,12 +140,18 @@ protected void channelRead0(final ChannelHandlerContext ctx, final HttpObject ms
 				}
 
 				if (httpContent instanceof LastHttpContent) {
+					LOG.trace("Finalizing multipart file upload.");
 					ctx.channel().attr(UPLOADED_FILES).set(new FileUploads(currentUploadDir));
-					ctx.fireChannelRead(currentHttpRequest);
 					if (currentJsonPayload != null) {
+						currentHttpRequest.headers().set(HttpHeaders.Names.CONTENT_LENGTH, currentJsonPayload.length);
+						currentHttpRequest.headers().set(HttpHeaders.Names.CONTENT_TYPE, RestConstants.REST_CONTENT_TYPE);
+						ctx.fireChannelRead(currentHttpRequest);
 						ctx.fireChannelRead(httpContent.replace(Unpooled.wrappedBuffer(currentJsonPayload)));
 					} else {
-						ctx.fireChannelRead(ReferenceCountUtil.retain(httpContent));
```

Comment:
Forwarding a single httpContent object that may or may not contain content always seemed fishy to me, so we now always send an `EMPTY_LAST_CONTENT` object instead.